### ISSUE TITLE
[HOLD for payment 2023-02-08] [$250] Strip workspace name of html or handle validation php side reported by @kerupuksambel

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12268]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open [staging.new.expensify.com](http://staging.new.expensify.com/)
2. Create new workspace
3. Change the name of the workspace with `<b> </b>` and save it

## Expected Result:
 Either the workspace name changed to <b> </b> after HTML-escaping (something like &lt;b&gt; &lt;/b&gt; ), or returned error  after the HTML tag got stripped and only the space remained

## Actual Result:
The workspace name changed to no name

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web

**Version Number:** 1.2.21-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/198730117-135e716c-ada7-4069-b153-df931cee5fc1.mp4

https://user-images.githubusercontent.com/43996225/198730150-69afdab6-b896-461c-945c-97c8914b4e39.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @kerupuksambel
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1666949886613589

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by NicMendonca** [comment]
I am having trouble reproducing this. Asked in slack [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1666949886613589).

**Comment by NicMendonca** [comment]
Okay, I got it. Am able to reproduce with using HTML spacing: 

![image](https://user-images.githubusercontent.com/24947345/199086995-f44e1b6f-fde3-4b93-bd3c-e21674277819.png)

**Comment by ctkochan22** [comment]
Looks related to https://github.com/Expensify/App/issues/12000

**Comment by trjExpensify** [comment]
Got it, so does it need to be internal as per this comment? https://github.com/Expensify/App/issues/12000#issuecomment-1296576574

**Comment by trjExpensify** [comment]
@ctkochan22 bump on the above question?

**Comment by s77rt** [comment]
### Proposal
```diff
diff --git a/src/pages/workspace/WorkspaceSettingsPage.js b/src/pages/workspace/WorkspaceSettingsPage.js
index e778ecf89..feefb74c7 100644
--- a/src/pages/workspace/WorkspaceSettingsPage.js
+++ b/src/pages/workspace/WorkspaceSettingsPage.js
@@ -56,7 +56,7 @@ class WorkspaceSettingsPage extends React.Component {
         if (this.props.policy.isPolicyUpdating) {
             return;
         }
-        const name = values.name.trim();
+        const name = values.name?.replace(/(<([^>]+)>)/ig, '').trim();
         const outputCurrency = values.currency;
         Policy.updateGeneralSettings(this.props.policy.id, name, outputCurrency);
         Keyboard.dismiss();
@@ -64,7 +64,8 @@ class WorkspaceSettingsPage extends React.Component {
 
     validate(values) {
         const errors = {};
-        if (!values.name || !values.name.trim().length) {
+        const name = values.name?.replace(/(<([^>]+)>)/ig, '').trim();
+        if (!name || !name.length) {
             errors.name = this.props.translate('workspace.editor.nameIsRequiredError');
         }
         return errors;

https://user-images.githubusercontent.com/16493223/200658772-e73798d4-fca0-4843-9ae9-3c46fb5073c5.mp4


```

**Comment by ctkochan22** [comment]
> Got it, so does it need to be internal as per this comment? https://github.com/Expensify/App/issues/12000#issuecomment-1296576574

I think i'm wrong about that. They seem like separate issues.

**Comment by s77rt** [comment]
@rushatgabhane 
1. Strip any html tags (using regex `(<([^>]+)>)/ig`
2. Trim the output
3. Check if the name is defined and has a length (>0)

**Comment by s77rt** [comment]
### Another Proposal (idea)
Instead of stripping the tags, we can encode them, so the user would have the ability to name the workspace something like `<test> WokrspaceA`. However the tags gets stripped in the API. As long as the API strips the tags (and not escape them) the first proposal is the best fit here.

**Comment by ctkochan22** [comment]
Seems like https://stackoverflow.com/questions/31515999/how-to-remove-all-html-tags-from-a-string has a better filtering. Strangely we don't do this anywhere else either. 

I think we should also validate the end result on the PHP side, and have a properly caught error.

**Comment by ctkochan22** [comment]
@s77rt thanks for your proposal. Let me try doing a PHP side error handler first and see how that goes

**Comment by s77rt** [comment]
### Updated Proposal
```diff
diff --git a/src/pages/workspace/WorkspaceSettingsPage.js b/src/pages/workspace/WorkspaceSettingsPage.js
index e778ecf89..a20413c68 100644
--- a/src/pages/workspace/WorkspaceSettingsPage.js
+++ b/src/pages/workspace/WorkspaceSettingsPage.js
@@ -56,7 +56,7 @@ class WorkspaceSettingsPage extends React.Component {
         if (this.props.policy.isPolicyUpdating) {
             return;
         }
-        const name = values.name.trim();
+        const name = values.name.replace(/<(.|\n)*?>/g, '').trim();
         const outputCurrency = values.currency;
         Policy.updateGeneralSettings(this.props.policy.id, name, outputCurrency);
         Keyboard.dismiss();
@@ -64,7 +64,7 @@ class WorkspaceSettingsPage extends React.Component {
 
     validate(values) {
         const errors = {};
-        if (!values.name || !values.name.trim().length) {
+        if (!values.name || !values.name.replace(/<(.|\n)*?>/g, '').trim().length) {
             errors.name = this.props.translate('workspace.editor.nameIsRequiredError');
         }
         return errors;
```
- The change is clearer this way
- The first proposal will return a literal "undefined" in case the name was undefined (due to using .replace with an undefined)
- Using the suggested regex

https://user-images.githubusercontent.com/16493223/200758101-0d5e4597-ac18-4983-96fe-618250f657cf.mp4

**Comment by trjExpensify** [comment]
@ctkochan22 does this need the `Internal` label? Also, it's a `Bug` in the App repo, so we should keep this on `Daily` as we look to achieve BugZero for WAQ. 

Can we confirm what the expected result is here and update the OP. Is this is? 

- If the workspace name includes HTML tags with no value between them (i.e `<b> <b>`, treat this as "blank" and show the error message that a workspace name is required.
- If the workspace name includes HTML tags with a value between them (i.e `<b> My workspace <b>`), display `My workspace` as the workspace name. ✅ - this is working as expected

FWIW, I thought I'd test this out on Profile > FirstName/LastName fields to compare the behaviour and how we handle it, but we have a fall back on the primaryLogin when a name isn't set: 

`<b> value <b>` ✅ (same scenario for the workspace name)
<img width="802" alt="image" src="https://user-images.githubusercontent.com/16232057/200815451-812464c7-6690-43ee-beec-4c668402fd33.png">

`<b> <b>`
blank (thus reverting to the primaryLogin, which is okay in this use-case because a name is optional). 
<img width="819" alt="image" src="https://user-images.githubusercontent.com/16232057/200815670-69159edd-0d5d-454d-97c6-027c6437a5d2.png">

Thanks!

**Comment by ctkochan22** [comment]
Sorry @trjExpensify going to go back on what I said, and bring this back internal.


My reasoning is as a rule of thumb, we always should have error catches as far server-side as possible and in each layer. Auth is ideal. Then PHP. Then client-side. 

I'm going to add a catch PHP side, and we'll return an error.

@trjExpensify How do you think we should handle this error? Do you happen to know what our convention is with errors we don't explicitly need users to "clear"

**Comment by trjExpensify** [comment]
Cool, switched the label. 


>@trjExpensify How do you think we should handle this error? Do you happen to know what our convention is with errors we don't explicitly need users to "clear"

Yep, let's follow the guidelines for form alerts: https://github.com/Expensify/App/blob/main/contributingGuides/FORMS.md#form-alerts

We largely have what we need here: 

<img width="372" alt="image" src="https://user-images.githubusercontent.com/16232057/200976069-524583ea-b6f6-4c7e-a798-a8e8add463f1.png">

**Comment by ctkochan22** [comment]
The error will look a little different because its an error being returned form the php side.

<img width="997" alt="image" src="https://user-images.githubusercontent.com/12980144/201234499-c99ab315-e6cd-4638-aea3-40a90698a08c.png">

Following how the code in the Form works at the moment.

**Comment by trjExpensify** [comment]
Got it. It would be better if that was in-line to the field in error really.

**Comment by ctkochan22** [comment]
Yeah, I'm trying to see if I can make that happen. Currently I only know how to do that as a pending error, meaning the user would have to click an "x" to get rid of it. Which I don't like. 

There must be a way

**Comment by trjExpensify** [comment]
When there's a will, there's a way. 😉

**Comment by ctkochan22** [comment]
So we are discussing how to do inline here: https://expensify.slack.com/archives/C03TQ48KC/p1668128270722689
- Seems like a new PR could allow us to add inline errors - https://github.com/Expensify/App/pull/11465

In the meantime, this is the Web PR to get it fixed: https://github.com/Expensify/Web-Expensify/pull/35548

**Comment by s77rt** [comment]
@rushatgabhane may be external as well (client side validation)

**Comment by trjExpensify** [comment]
Okay awesome, so @ctkochan22 are we holding that PR on @youssef-lr's [here](https://github.com/Expensify/App/pull/11465)?

**Comment by youssef-lr** [comment]
My PR will be ready today. But I think we should also add client side validation for this.

**Comment by trjExpensify** [comment]
👋 What's the latest on that PR, @youssef-lr @ctkochan22? It looks like it has conflicts now and its been in review for a week.

**Comment by trjExpensify** [comment]
What's the latest on that one now as the issue has closed?

**Comment by youssef-lr** [comment]
@trjExpensify I just merged @ctkochan22's Web PR. I think we'll still need to handle validation in the frontend as well before we can close this out.

**Comment by trjExpensify** [comment]
Awesome. Thanks for the update!

**Comment by trjExpensify** [comment]
Okay, what's the next step here? Who's taking the PR on the front-end validation if that's deemed a requirement to close this one out?

**Comment by trjExpensify** [comment]
Bump @youssef-lr @ctkochan22?

**Comment by ctkochan22** [comment]
Was ooo last week for a family emergency. I can do it

**Comment by youssef-lr** [comment]
@trjExpensify I think what's left to implement here is stripping the name from HTML tags then verifying if it's empty, we're doing this in the backend and it needs to be done in the client as well. I think @ctkochan22 might have already implemented this and hasn't made the PR yet. The linked issue only shows validation of a completely empty name, e.g. `' '` vs `'<b></b>'` (both of these should be considered empty).

**Comment by trjExpensify** [comment]
Okay, I'm still slightly confused by that linked issue if it's not addressing the client-side validation follow-up required for this issue. I'm sure @ctkochan22 can confirm!

**Comment by trjExpensify** [comment]
>I think @ctkochan22 might have already implemented this and hasn't made the PR yet.

@ctkochan22 can you confirm on this please?

**Comment by ctkochan22** [comment]
Yes, been playing catch up after 3 unexpected weeks of OOO. I'll have a PR out today for sure.

**Comment by trjExpensify** [comment]
Thanks, @ctkochan22! Is the PR up to link here?

**Comment by ctkochan22** [comment]
So currently this is what happens when you try to submit an html encoded name that results in spaces:
<img width="354" alt="image" src="https://user-images.githubusercontent.com/12980144/210496182-08e663e0-032e-4983-b1fd-9eb095d34339.png">

We might need to update the copy?

**Comment by trjExpensify** [comment]
Yeah, it might be a tad confusing as the user thinks they have defined a name don't they?

**Comment by trjExpensify** [comment]
@ctkochan22 how you getting on here?

**Comment by ctkochan22** [comment]
`"Please enter a valid workspace name"`

Just need to fix the copy I think?

**Comment by trjExpensify** [comment]
What's the validation exactly? We don't allow HTML tags, or we do, but it's because this is effectively blank?

**Comment by ctkochan22** [comment]
We strip html tags out when it goes to the php. I wonder if we want to allow that however?

**Comment by ctkochan22** [comment]
Sorry, trying to get another WAQ issue out too. 

If we want to set html tags as invalid, we may need to change our approach. Because right now, 
1. Set `<b> </b>` as workspace
2. WAF input rules filters all html out. So when it hits html, its just a space. 
3. On the php side, we [now prevent adding an empty space ](https://github.com/Expensify/Web-Expensify/pull/35548)as a workspace name. 

To validate and prevent submitting html tags onBlur, instead of using `.replace` [here](https://github.com/Expensify/App/pull/13978/files#diff-97f2ba8101f628a954551199fd111b5a5aa30d2a744829c85fb993cd8700cebdR75). We could use `name.search(/<(.|\n)*?>/g)` to check if it has html tags in it (anything with `<string>`.

I'd also like to float out there that we allow policies in oldDot to be named with html tags in it. Are we sure we don't want to allow them to save workspace names with html tags in them?

**Comment by trjExpensify** [comment]
I prefer not to allow HTML tags and use the established Forms pattern to validate onBlur. Reasons being:

- I can't really see a case for where you need to use HTML tags in a workspace name
- They're not doing anything i.e `<b> Tom's Workspace <b>` isn't being displayed as "**Tom's Workspace**" everywhere the workspaceName features throughout the App

**Comment by JmillsExpensify** [comment]
Happened to come across this one scanning through the repo. I think I'm in a similar spot. I don't see why we'd allow tags, though I bet it stems from either inconsistent or a different philosophy on inputs/form validation in general. Given that they do nothing, I'd just as well say we shouldn't allow HTML tags.

**Comment by ctkochan22** [comment]
Screenshot
<img width="345" alt="image" src="https://user-images.githubusercontent.com/12980144/213584894-5479f8b1-f393-4fdd-ae0a-395e994435a7.png">

**Comment by ctkochan22** [comment]
If the error message and screenshot looks fine, you can hop over to the PR to check out the video. https://github.com/Expensify/App/pull/13978

**Comment by ctkochan22** [comment]
During code review @flodnv asked whether we should add this validation to other form/fields as well?

https://github.com/Expensify/App/pull/13978#discussion_r1084123725

I would rather just get this out now as is, and people can add this validation to other fields in need be. However, to @flodnv 's argument, there are probably a lot of other fields, where if you put in html tags. We strip it out on the PHP side with no pre-validation on the App side.

**Comment by ctkochan22** [comment]
Like in display name. If you put in a `<b> </b>` as the first name, it fails silently. This is probably because the tags, items wrapped in `< >` are stripped out before reaching PHP. So then we try to save the first name with an empty space. Which is probably not allowed in PHP.

**Comment by flodnv** [comment]
I think we should fix this holistically. There is no point in fixing this 42 times now, and 42 times in 3 years from now after we've added hundreds of inputs that don't allow HTML.

**Comment by JmillsExpensify** [comment]
I agree with that, though I also think we can create a separate issue to do that too, listing out every place where we need to address the inputs/HTML. Like given that this started and as a bug report, I think it's fair to fix this bug and not close this issue until another more holistic one is created.

**Comment by flodnv** [comment]
I'm ok with that, but I also don't see the reason to do it this way 😄

**Comment by ctkochan22** [comment]
Lets create two GH's. One to address this one case, that required a [Web-Expensify PR](https://github.com/Expensify/Web-Expensify/pull/35548) and this App PR. Then we can close this out completely.

The other GH (that I can create and assign to myself) will list all the other cases we want this validation. Also checks whether we need PHP changes. And will hopefully be easier to list out all the QA we need to do on each of those forms?

**Comment by flodnv** [comment]
Sure, if you want.

>list all the other cases we want this validation.

What I am saying is that all fields should disallow HTML. I don't know of any field that should allow it.

**Comment by trjExpensify** [comment]
Yeah, I'm fine going that route @ctkochan22 if you can take it 👍

**Comment by ctkochan22** [comment]
> What I am saying is that all fields should disallow HTML. I don't know of any field that should allow it.

I understand. My issue is we will need to compile all the cases, add php checks, and QA appropriately. I think it'll be cleaner in another issue. 

Created the other github: https://github.com/Expensify/App/issues/14610

**Comment by flodnv** [comment]
Thanks @ctkochan22

**Comment by ctkochan22** [comment]
> the regression test, given that adding HTML tags to a workspace name is very unlikely in real-life use, I don't think we add a test case for it to execute on every regression run.

I agree.

**Comment by trjExpensify** [comment]
Awesome, thanks for confirming. @kerupuksambel I've sent you an offer in Upwork.

**Comment by trjExpensify** [comment]
Thanks, settled up!
